File: closetalk_backend/internal/model/message.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Message struct {
	ID               uuid.UUID   `json:"id"`
	ChatID           string      `json:"chat_id"`
	SenderID         string      `json:"sender_id"`
	SenderDeviceID   string      `json:"sender_device_id,omitempty"`
	RecipientIDs     []string    `json:"recipient_ids,omitempty"` // for multi-device fan-out
	Content          string      `json:"content"`
	ContentType      string      `json:"content_type"` // text | image | video | file | voice | poll
	MediaURL         string      `json:"media_url,omitempty"`
	MediaID          string      `json:"media_id,omitempty"`
	ReplyToID        *uuid.UUID  `json:"reply_to_id,omitempty"`
	ForwardedFrom    string      `json:"forwarded_from,omitempty"`
	Status           string      `json:"status"` // sending | sent | delivered | read
	ModerationStatus string      `json:"moderation_status,omitempty"`
	EditHistory      []EditEntry `json:"edit_history,omitempty"`
	IsDeleted        bool        `json:"is_deleted,omitempty"`
	CreatedAt        time.Time   `json:"created_at"`
	EditedAt         *time.Time  `json:"edited_at,omitempty"`
	DisappearedAt    *time.Time  `json:"disappeared_at,omitempty"`
}

type EditEntry struct {
	Content  string    `json:"content"`
	EditedAt time.Time `json:"edited_at"`
}

type SendMessageRequest struct {
	ChatID         string   `json:"chat_id"`
	Content        string   `json:"content"`
	ContentType    string   `json:"content_type"`
	MediaURL       string   `json:"media_url,omitempty"`
	MediaID        string   `json:"media_id,omitempty"`
	ReplyToID      string   `json:"reply_to_id,omitempty"`
	ForwardedFrom  string   `json:"forwarded_from,omitempty"`
	DisappearAfter string   `json:"disappear_after,omitempty"`
	RecipientIDs   []string `json:"recipient_ids,omitempty"` // for multi-device fan-out
}

type ForwardMessageRequest struct {
	MessageID     string   `json:"message_id"`
	TargetChatIDs []string `json:"target_chat_ids"`
}

type EditMessageRequest struct {
	Content string `json:"content"`
}

type ReactToMessageRequest struct {
	Emoji string `json:"emoji"`
}

type MessageResponse struct {
	ID               uuid.UUID   `json:"id"`
	ChatID           string      `json:"chat_id"`
	SenderID         string      `json:"sender_id"`
	SenderUsername   string      `json:"sender_username,omitempty"`
	RecipientIDs     []string    `json:"recipient_ids,omitempty"`
	Content          string      `json:"content"`
	ContentType      string      `json:"content_type"`
	MediaURL         string      `json:"media_url,omitempty"`
	MediaID          string      `json:"media_id,omitempty"`
	ReplyToID        *uuid.UUID  `json:"reply_to_id,omitempty"`
	ForwardedFrom    string      `json:"forwarded_from,omitempty"`
	Status           string      `json:"status"`
	ModerationStatus string      `json:"moderation_status,omitempty"`
	EditHistory      []EditEntry `json:"edit_history,omitempty"`
	IsDeleted        bool        `json:"is_deleted,omitempty"`
	Reactions        []Reaction  `json:"reactions,omitempty"`
	CreatedAt        time.Time   `json:"created_at"`
	EditedAt         *time.Time  `json:"edited_at,omitempty"`
	DisappearedAt    *time.Time  `json:"disappeared_at,omitempty"`
}

type Reaction struct {
	UserID    string    `json:"user_id"`
	Emoji     string    `json:"emoji"`
	CreatedAt time.Time `json:"created_at"`
}

type PaginatedMessages struct {
	Messages   []MessageResponse `json:"messages"`
	NextCursor string            `json:"next_cursor,omitempty"`
	HasMore    bool              `json:"has_more"`
}

type BookmarkRequest struct {
	MessageID string `json:"message_id"`
	ChatID    string `json:"chat_id"`
}

type BookmarkResponse struct {
	MessageID string    `json:"message_id"`
	ChatID    string    `json:"chat_id"`
	Preview   string    `json:"content_preview"`
	CreatedAt time.Time `json:"created_at"`
}

type WebSocketMessage struct {
	Type    string `json:"type"` // message.new | message.updated | message.status | typing.start | typing.stop
	Payload any    `json:"payload"`
}

type ReadReceipt struct {
	MessageID string    `json:"message_id"`
	UserID    string    `json:"user_id"`
	ReadAt    time.Time `json:"read_at"`
}

type SyncMessagesRequest struct {
	After string `json:"after,omitempty"` // cursor (created_at timestamp)
	Limit int    `json:"limit,omitempty"`
}

type SyncMessagesResponse struct {
	Messages   []MessageResponse `json:"messages"`
	NextCursor string            `json:"next_cursor,omitempty"`
	HasMore    bool              `json:"has_more"`
}

type SyncStatusRequest struct {
	After string `json:"after,omitempty"`
}

type SyncStatusResponse struct {
	Statuses   []StatusEntry `json:"statuses"`
	NextCursor string        `json:"next_cursor,omitempty"`
	HasMore    bool          `json:"has_more"`
}

type StatusEntry struct {
	ID        string    `json:"id"`
	UserID    string    `json:"user_id"`
	Type      string    `json:"type"` // text | image | video
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
	ExpiresAt time.Time `json:"expires_at"`
}

type SearchMessagesRequest struct {
	ChatID string `json:"chat_id"`
	Query  string `json:"query"`
	Cursor string `json:"cursor,omitempty"`
	Limit  int    `json:"limit,omitempty"`
}

type SearchResult struct {
	MessageID   string    `json:"message_id"`
	ChatID      string    `json:"chat_id"`
	SenderID    string    `json:"sender_id"`
	SenderName  string    `json:"sender_name,omitempty"`
	Content     string    `json:"content"`
	ContentType string    `json:"content_type"`
	Snippet     string    `json:"snippet"`
	CreatedAt   time.Time `json:"created_at"`
}

type SearchMessagesResponse struct {
	Results    []SearchResult `json:"results"`
	NextCursor string         `json:"next_cursor,omitempty"`
	HasMore    bool           `json:"has_more"`
}

type SearchMessagesRequestParams struct {
	Query  string `json:"q"`
	Cursor string `json:"cursor"`
	Limit  int    `json:"limit"`
}

File: closetalk_app/lib/models/message.dart
class Message {
  final String id;
  final String chatId;
  final String senderId;
  final String? senderUsername;
  final String content;
  final String contentType;
  final String? mediaUrl;
  final String? mediaId;
  final String? replyToId;
  final String? forwardedFrom;
  final String status;
  final String? moderationStatus;
  final bool isDeleted;
  final List<Reaction> reactions;
  final DateTime createdAt;
  final DateTime? editedAt;
  final DateTime? disappearedAt;

  Message({
    required this.id,
    required this.chatId,
    required this.senderId,
    this.senderUsername,
    required this.content,
    this.contentType = 'text',
    this.mediaUrl,
    this.mediaId,
    this.replyToId,
    this.forwardedFrom,
    this.status = 'sending',
    this.moderationStatus,
    this.isDeleted = false,
    this.reactions = const [],
    required this.createdAt,
    this.editedAt,
    this.disappearedAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as String,
      chatId: json['chat_id'] as String,
      senderId: json['sender_id'] as String,
      senderUsername: json['sender_username'] as String?,
      content: json['content'] as String,
      contentType: json['content_type'] as String? ?? 'text',
      mediaUrl: json['media_url'] as String?,
      mediaId: json['media_id'] as String?,
      replyToId: json['reply_to_id'] as String?,
      forwardedFrom: json['forwarded_from'] as String?,
      status: json['status'] as String? ?? 'sent',
      moderationStatus: json['moderation_status'] as String?,
      isDeleted: json['is_deleted'] as bool? ?? false,
      reactions: (json['reactions'] as List<dynamic>?)
              ?.map((e) => Reaction.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at'] as String),
      editedAt: json['edited_at'] != null
          ? DateTime.parse(json['edited_at'] as String)
          : null,
      disappearedAt: json['disappeared_at'] != null
          ? DateTime.parse(json['disappeared_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'chat_id': chatId,
        'sender_id': senderId,
        'sender_username': senderUsername,
        'content': content,
        'content_type': contentType,
        'media_url': mediaUrl,
        'media_id': mediaId,
        'reply_to_id': replyToId,
        'forwarded_from': forwardedFrom,
        'status': status,
        'moderation_status': moderationStatus,
        'is_deleted': isDeleted,
        'reactions': reactions.map((r) => r.toJson()).toList(),
        'created_at': createdAt.toIso8601String(),
        'edited_at': editedAt?.toIso8601String(),
        'disappeared_at': disappearedAt?.toIso8601String(),
      };

  Message copyWith({
    String? id,
    String? chatId,
    String? senderId,
    String? senderUsername,
    String? content,
    String? contentType,
    String? mediaUrl,
    String? mediaId,
    String? replyToId,
    String? forwardedFrom,
    String? status,
    String? moderationStatus,
    bool? isDeleted,
    List<Reaction>? reactions,
    DateTime? createdAt,
    DateTime? editedAt,
    DateTime? disappearedAt,
  }) {
    return Message(
      id: id ?? this.id,
      chatId: chatId ?? this.chatId,
      senderId: senderId ?? this.senderId,
      senderUsername: senderUsername ?? this.senderUsername,
      content: content ?? this.content,
      contentType: contentType ?? this.contentType,
      mediaUrl: mediaUrl ?? this.mediaUrl,
      mediaId: mediaId ?? this.mediaId,
      replyToId: replyToId ?? this.replyToId,
      forwardedFrom: forwardedFrom ?? this.forwardedFrom,
      status: status ?? this.status,
      moderationStatus: moderationStatus ?? this.moderationStatus,
      isDeleted: isDeleted ?? this.isDeleted,
      reactions: reactions ?? this.reactions,
      createdAt: createdAt ?? this.createdAt,
      editedAt: editedAt ?? this.editedAt,
      disappearedAt: disappearedAt ?? this.disappearedAt,
    );
  }
}

class Reaction {
  final String userId;
  final String emoji;
  final DateTime createdAt;

  Reaction({
    required this.userId,
    required this.emoji,
    required this.createdAt,
  });

  factory Reaction.fromJson(Map<String, dynamic> json) {
    return Reaction(
      userId: json['user_id'] as String,
      emoji: json['emoji'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'user_id': userId,
        'emoji': emoji,
        'created_at': createdAt.toIso8601String(),
      };
}

class SearchResult {
  final String messageId;
  final String chatId;
  final String senderId;
  final String senderName;
  final String content;
  final String contentType;
  final String snippet;
  final DateTime createdAt;

  SearchResult({
    required this.messageId,
    required this.chatId,
    required this.senderId,
    required this.senderName,
    required this.content,
    required this.contentType,
    required this.snippet,
    required this.createdAt,
  });

  factory SearchResult.fromJson(Map<String, dynamic> json) {
    return SearchResult(
      messageId: json['message_id'] as String,
      chatId: json['chat_id'] as String,
      senderId: json['sender_id'] as String,
      senderName: json['sender_name'] as String? ?? '',
      content: json['content'] as String,
      contentType: json['content_type'] as String? ?? 'text',
      snippet: json['snippet'] as String? ?? '',
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

class SearchMessagesResponse {
  final List<SearchResult> results;
  final String? nextCursor;
  final bool hasMore;

  SearchMessagesResponse({
    required this.results,
    this.nextCursor,
    this.hasMore = false,
  });

  factory SearchMessagesResponse.fromJson(Map<String, dynamic> json) {
    return SearchMessagesResponse(
      results: (json['results'] as List<dynamic>?)
              ?.map((e) => SearchResult.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      nextCursor: json['next_cursor'] as String?,
      hasMore: json['has_more'] as bool? ?? false,
    );
  }
}

class PaginatedMessages {
  final List<Message> messages;
  final String? nextCursor;
  final bool hasMore;

  PaginatedMessages({
    required this.messages,
    this.nextCursor,
    this.hasMore = false,
  });

  factory PaginatedMessages.fromJson(Map<String, dynamic> json) {
    return PaginatedMessages(
      messages: (json['messages'] as List<dynamic>)
          .map((e) => Message.fromJson(e as Map<String, dynamic>))
          .toList(),
      nextCursor: json['next_cursor'] as String?,
      hasMore: json['has_more'] as bool? ?? false,
    );
  }
}
